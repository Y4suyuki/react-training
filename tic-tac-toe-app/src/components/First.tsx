import React from 'react';
import { Player } from '../types/Player';

export type FirstProps = {
  setPlayer: (player: Player) => void;
};

export const First: React.FC<FirstProps> = ({setPlayer}) => {
  return (
    <>
      <div>
        Select a player.
      </div>
      <button onClick={() => setPlayer('X')} style={{marginTop: "10px", marginRight: "10px"}}>
        X
      </button>
      <button onClick={() => setPlayer('O')}>
        O
      </button>
    </>
  );
};
