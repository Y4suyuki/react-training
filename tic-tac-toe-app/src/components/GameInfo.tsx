import { History, Mark } from "../types";

type Props = {
  winner: Mark;
  xIsNext: boolean;
  history: History;
  jumpTo: (i: number) => void;
};

function MoveInfo(props: { onClick: () => void; move: number }) {
  const desc = props.move ? "Go to move #" + props.move : "Go to game start";
  return (
    <li>
      <button onClick={props.onClick}>{desc}</button>
    </li>
  );
}

export function GameInfo(props: Props) {
  let status: string;
  if (props.winner) {
    status = "Winner: " + props.winner;
  } else {
    status = "Next player: " + (props.xIsNext ? "X" : "O");
  }

  return (
    <div className="game-info">
      <div>{status}</div>
      <ol>
        {props.history.map((_, i) => {
          return <MoveInfo onClick={() => props.jumpTo(i)} move={i} />;
        })}
      </ol>
    </div>
  );
}
