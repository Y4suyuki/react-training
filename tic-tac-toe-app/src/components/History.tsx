import React from 'react';
import { Player } from '../types/Player'

type HistoryProps = {
  xIsNext: boolean,
  moveHistory: {
    squares: Player[]
  }[],
  winner: Player,
  jumpTo: (move: number) => void
}

export const History: React.FC<HistoryProps> = ({xIsNext, moveHistory, winner, jumpTo}) => {
  const status =
    winner
      ? 'Winner: ' +  winner
      : 'Next player: ' + (xIsNext ? 'X' : 'O');

  const moves = moveHistory.map((_, move) => {
    const desc = move ?
      'Go to move #' + move :
      'Go to game start';

    return (
      <li key={move}>
        <button onClick={() => jumpTo(move)}>{desc}</button>
      </li>
    );
  });

  return(
    <>
      <div>{status}</div>
      <ol>{moves}</ol>
    </>
  )
}
