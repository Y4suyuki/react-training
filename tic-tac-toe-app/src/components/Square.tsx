import { useEffect } from "react";
import { infoStyle, infoStyleRed } from "../constants";
import { Mark } from "../types";

type SquareProps = { value: Mark; onClick: () => void };

export function Square(props: SquareProps) {
  useEffect(() => {
    console.log(`%cupdating square`, infoStyle);
    return () => {
      console.log(`%cunmounting square`, infoStyleRed);
    };
  }, []);
  return (
    <button className="square" onClick={props.onClick}>
      {props.value}
    </button>
  );
}
