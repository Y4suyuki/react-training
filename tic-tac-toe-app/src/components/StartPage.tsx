import { useState } from "react";
import { choosePlayerSide } from "../helper";
import { Mark } from "../types";

export function StartPage(props: {
  startGame: () => void;
  setPlayerSide: (playerSide: Mark) => void;
}) {
  const [tempSide, setTempSide] = useState<Mark | "random">("X");

  function SideCheckBox(props: { choice: Mark | "random" }) {
    return (
      <div>
        <input
          type="checkbox"
          checked={tempSide === props.choice}
          onChange={() => setTempSide(props.choice)}
        />
        <label>{props.choice}</label>
      </div>
    );
  }
  return (
    <div className="start-page">
      <h1>start game</h1>
      <h2>select your side</h2>
      <div className="side-selector">
        <SideCheckBox choice="X" />
        <SideCheckBox choice="O" />
        <SideCheckBox choice="random" />
      </div>
      <button
        className="start-button"
        onClick={() => {
          if (tempSide === "random") {
            props.setPlayerSide(choosePlayerSide());
          } else {
            props.setPlayerSide(tempSide);
          }
          props.startGame();
        }}
      >
        start game
      </button>
    </div>
  );
}
