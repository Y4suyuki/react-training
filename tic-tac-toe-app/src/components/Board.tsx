import React from 'react';
import { SquareValue } from '../types/SquareValue'
import { Square } from '../components/Square'

type BordProps = {
  squares: SquareValue[];
  onClick: (i :number) => void;
};

export class Board extends React.Component<BordProps> {
  renderSquare(i: number) {
    return (
      <Square
        value={this.props.squares[i]}
        onClick={() => this.props.onClick(i)}
      />
    );
  }

  render() {
    return (
      <div>
        <div className="board-row">
          {this.renderSquare(0)}
          {this.renderSquare(1)}
          {this.renderSquare(2)}
        </div>
        <div className="board-row">
          {this.renderSquare(3)}
          {this.renderSquare(4)}
          {this.renderSquare(5)}
        </div>
        <div className="board-row">
          {this.renderSquare(6)}
          {this.renderSquare(7)}
          {this.renderSquare(8)}
        </div>
      </div>
    );
  }
};
