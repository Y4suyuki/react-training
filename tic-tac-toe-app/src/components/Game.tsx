import { Board } from '../components/Board'
import { useGame } from '../hooks/useGame'
import { calculateWinner } from '../lib/calculateWinner'

export const Game: React.FC = () => {
  const { history, stepNumber, xIsNext, handleClick, jumpTo } = useGame();

  const current = history[stepNumber];
  const winner = calculateWinner(current.squares);
  const status =
    winner
      ? 'Winner: ' +  winner
      : 'Next player: ' + (xIsNext ? 'X' : 'O');
  const moves = history.map((step, move) => {
    const desc = move ?
      'Go to move #' + move :
      'Go to game start';

    return (
      <li key={move}>
        <button onClick={() => jumpTo(move)}>{desc}</button>
      </li>
    );
  });

  return (
    <div className="game">
      <div className="game-board">
        <Board
          squares={current.squares}
          onClick={(i) => handleClick(i)}
        />
      </div>
      <div className="game-info">
        <div>{status}</div>
        <ol>{moves}</ol>
      </div>
    </div>
  );
};
