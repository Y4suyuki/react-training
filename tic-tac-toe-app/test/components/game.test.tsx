import { Game } from '../../src/components/game'
import { render } from '@testing-library/react'

describe('Game', () => {
  const game = render(<Game />)
  test('snapshot', () => {
    expect(game).toMatchSnapshot()
  })
})
