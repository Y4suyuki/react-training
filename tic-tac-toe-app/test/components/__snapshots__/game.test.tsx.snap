// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="game"
      >
        <div
          class="game-board"
        >
          <div>
            <div
              class="board-row"
            >
              <button
                class="square"
              />
              <button
                class="square"
              />
              <button
                class="square"
              />
            </div>
            <div
              class="board-row"
            >
              <button
                class="square"
              />
              <button
                class="square"
              />
              <button
                class="square"
              />
            </div>
            <div
              class="board-row"
            >
              <button
                class="square"
              />
              <button
                class="square"
              />
              <button
                class="square"
              />
            </div>
          </div>
        </div>
        <div
          class="game-info"
        >
          <div>
            Next player: X
          </div>
          <ol>
            <li>
              <button>
                Go to game start
              </button>
            </li>
          </ol>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="game"
    >
      <div
        class="game-board"
      >
        <div>
          <div
            class="board-row"
          >
            <button
              class="square"
            />
            <button
              class="square"
            />
            <button
              class="square"
            />
          </div>
          <div
            class="board-row"
          >
            <button
              class="square"
            />
            <button
              class="square"
            />
            <button
              class="square"
            />
          </div>
          <div
            class="board-row"
          >
            <button
              class="square"
            />
            <button
              class="square"
            />
            <button
              class="square"
            />
          </div>
        </div>
      </div>
      <div
        class="game-info"
      >
        <div>
          Next player: X
        </div>
        <ol>
          <li>
            <button>
              Go to game start
            </button>
          </li>
        </ol>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
